### PR TITLE
feat(reporters): `summary` option for `verbose` and `default` reporters

### PR DESCRIPTION
**Pull request by AriPerkkio**
### Description

- Closes https://github.com/vitest-dev/vitest/issues/3732
- Fixes https://github.com/vitest-dev/vitest/issues/6590
- Fixes https://github.com/vitest-dev/vitest/issues/6177
- Requires https://github.com/tinylibs/tinypool/pull/104

Adds new `summary: boolean` option for `verbose` and `default` reporters. This option enables test summary that is shown below test results. This option is enabled by default. I would not consider this as breaking change, rather as bug fix for issues linked above.

```ts
import { defineConfig } from 'vitest/config'

export default defineConfig({
  test: {
    reporters: [
      ['verbose', { summary: true }],

      // or
      ['default', { summary: true }],
    ],
  },
})
```

Running Vitest's `test/core` tests - left one is `main`, right one is this PR:

[vitest-reporter-before-and-after.webm](https://github.com/user-attachments/assets/59f374b9-2f55-46c4-a8b2-4cd1dd5d0a6e)

Most of the changes are related to `listRenderer.ts` removal. This is replaced with `windowedRenderer.ts`. After this PR we should do similar changes for `dotRenderer` and `tableRenderer`, and remove whole `log-update` from Vitest core.

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. If the feature is substantial or introduces breaking changes without a discussion, PR might be closed.
- [x] Ideally, include a test that fails without this PR but passes with it.
- [x] Please, don't make changes to `pnpm-lock.yaml` unless you introduce a new test example.

### Tests
- [x] Run the tests with `pnpm test:ci`.

### Documentation
- [x] If you introduce new functionality, document it. You can run documentation with `pnpm run docs` command.

### Changesets
- [x] Changes in changelog are generated from PR name. Please, make sure that it explains your changes in an understandable manner. Please, prefix changeset messages with `feat:`, `fix:`, `perf:`, `docs:`, or `chore:`.

### CODE REVIEW

File: packages/vitest/package.json [hunk 1]
```diff
@@ -165,7 +165,7 @@
     "std-env": "^3.7.0",
     "tinybench": "^2.9.0",
     "tinyexec": "^0.3.1",
-    "tinypool": "^1.0.1",
+    "tinypool": "https://pkg.pr.new/tinypool@67f3129",
```
Comment (AriPerkkio) [inline packages/vitest/package.json:168]: Update after releasing https://github.com/tinylibs/tinypool/pull/104

File: docs/guide/reporters.md [hunk 2]
```diff
@@ -96,33 +96,54 @@ This example will write separate JSON and XML reports as well as printing a verb
...
+ Test Files  4 passed (4)
+      Tests  16 passed | 4 skipped (20)
    Start at  12:34:32
    Duration  1.26s (transform 35ms, setup 1ms, collect 90ms, tests 1.47s, environment 0ms, prepare 267ms)
 ```
 
 ### Basic Reporter
 
-The `basic` reporter displays the test files that have run and a summary of results after the entire suite has finished running. Individual tests are not included in the report unless they fail.
+The `basic` reporter is equivalent to `default` reporter without `summary`.
```
Comment (sheremet-va) [inline docs/guide/reporters.md:146]: should we deprecate it?

File: docs/guide/reporters.md [hunk 3]
```diff
@@ -151,7 +172,7 @@ Example output using basic reporter:
 
 ### Verbose Reporter
 
-Follows the same hierarchical structure as the `default` reporter, but does not collapse sub-trees for passed test suites. The final terminal output displays all tests that have run, including those that have passed.
+Verbose reporter is same as `default` reporter, but it also displays each individual test after the suite has finished. Similar to `default` reporter, you can disable the summary by configuring the reporter.
```
Comment (sheremet-va) [inline docs/guide/reporters.md:175]: hm, it is quite useful to show currently running tests especially when they timeout

Comment (AriPerkkio) [reply docs/guide/reporters.md:175]: Yep I agree. However showing all tests like current `verbose` reporter does causes a lot of flickering as there is simply too much data that changes too fast.

How about we show names of running tests that have been running longer than `slowTestThreshold`?

Then we could see slow tests like:



And if there are other fast running tests, they will not expand:

Comment (sheremet-va) [reply docs/guide/reporters.md]: sure, seems like a good middle ground 👌
but don't forget they can run in parallel, so maybe we limit the number of tests shown?
and also beforeAll/beforeEach can take time (we previously were showing it)

Comment (AriPerkkio) [reply docs/guide/reporters.md]: Now tracking slow hooks too. The `onTaskUpdate` is tricky 😬 

```ts
import { afterAll, afterEach, beforeAll, beforeEach, test } from 'vitest'

beforeEach(async () => {
  await new Promise(resolve => setTimeout(resolve, 3_000))
})

beforeAll(async () => {
  await new Promise(resolve => setTimeout(resolve, 3_000))
})

afterEach(async () => {
  await new Promise(resolve => setTimeout(resolve, 3_000))
})

afterAll(async () => {
  await new Promise(resolve => setTimeout(resolve, 3_000))
})

test('fixture 1', async () => {
  await new Promise(resolve => setTimeout(resolve, 3_000))
})
```

[summary-hooks.webm](https://github.com/user-attachments/assets/6a4f1bb0-d740-4cfb-b1d0-2f726b586905)

Comment (sheremet-va) [reply docs/guide/reporters.md]: Oh wow, looks awesome!

**Comment by sheremet-va** [comment]
Merged PR that affects the output: https://github.com/vitest-dev/vitest/pull/6805 (it now prints a `result.note` message instead of `[skip]`)

File: packages/vitest/package.json (hunk 1 above)
Comment (AriPerkkio) [reply packages/vitest/package.json]: ✅ https://github.com/tinylibs/tinypool/releases/tag/v1.0.2

File: docs/guide/reporters.md (hunk 2 above)
Comment (sheremet-va) [reply docs/guide/reporters.md:146]: I think we can deprecate it straight away

Comment (AriPerkkio) [reply docs/guide/reporters.md:146]: Added

File: packages/vitest/src/node/reporters/renderers/windowedRenderer.ts [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+    for (let i = 1; i < this.windowHeight; i++) {
+      this.write(`${MOVE_CURSOR_ONE_ROW_UP}${CLEAR_LINE}`)
+    }
+
+    this.windowHeight = 0
+  }
+
+  private interceptStream(stream: NodeJS.WriteStream, type: StreamType) {
+    // @ts-expect-error -- not sure how 2 overloads should be typed
+    stream.write = (chunk, _, callback) => {
```
[the remark this answers is not in the document]
Comment (AriPerkkio) [reply packages/vitest/src/node/reporters/renderers/windowedRenderer.ts:149]: Good point, maybe it should. 

One case that comes to mind is that user calls `startVitest()` with custom stream. Once tests have finished and they've called `.close()`, we should restore the `process.stdout` back to it's orinal and not leave it pointing to the custom stream that was passed.

**Comment by saitonakamura** [comment]
Not sure if it's the best place to report it, but I think one thing was list during the migration: reporting of retried tests. list renderer contained this line 
https://github.com/vitest-dev/vitest/blob/171041af5e4f9e59eb4a446b7561cbfdc9da3c65/packages/vitest/src/node/reporters/renderers/listRenderer.ts#L138
it might be intentional though

**Pull request merged**